src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -427,11 +427,14 @@ abstract class RefChecks extends Transform {
                                           "(this rule is designed to prevent accidental overrides)")
             } else if (other.isStable && !member.isStable) { // (1.4)
               overrideErrorWithMemberInfo("stable, immutable value required to override:")
-            } else if (member.isValue && member.isLazy &&
+            } else if (member.isValue && (member.isLazy || member.isModule) &&
                        other.isValue && other.hasFlag(STABLE) && !(other.isDeferred || other.isLazy)) {
-              overrideErrorWithMemberInfo("concrete non-lazy value cannot be overridden:")
+              if (member.isModule && !currentRun.isScala3)
+                runReporting.deprecationWarning(member.pos, member, member.owner, "overriding concrete non-lazy value with an object member will be disallowed under -Xsource:3", "2.13.11")
```

Comment:
Hmm, it seems this comes too early now and prevents override checking for compatible types:

```
scala> class A { val value = 0 }
class A

scala> object O extends A { override object value }
                                            ^
       warning: overriding concrete non-lazy value with an object member will be disallowed under -Xsource:3
object O

scala> (O: A).value
java.lang.ClassCastException: class O$value$ cannot be cast to class java.lang.Integer (O$value$ is in unnamed module of loader scala.tools.nsc.interpreter.IMain$TranslatingClassLoader @475646d4; java.lang.Integer is in module java.base of loader 'bootstrap')
```

Reply:
Sorry, I mean to revisit. I think I said I'd break the if-else.